src/builder.rs

```diff
@@ -518,22 +551,58 @@ fn append_fs(
     append(dst, &header, read)
 }
 
+#[cfg(unix)]
+fn append_fs_special(
+    dst: &mut dyn Write,
+    path: &Path,
+    meta: &fs::Metadata,
+    entry_type: EntryType,
+    mode: HeaderMode,
+) -> io::Result<()> {
+    use ::std::os::unix::fs::MetadataExt;
+    let mut header = Header::new_gnu();
+    header.set_metadata_in_mode(meta, mode);
+    prepare_header_path(dst, &mut header, path)?;
+
+    header.set_entry_type(entry_type);
+    let dev_id = meta.rdev();
+    let dev_major = ((dev_id >> 32) & 0xffff_f000) | ((dev_id >> 8) & 0x0000_0fff);
+    let dev_minor = ((dev_id >> 12) & 0xffff_ff00) | ((dev_id) & 0x0000_00ff);
+    header.set_device_major(dev_major as u32)?;
+    header.set_device_minor(dev_minor as u32)?;
+
+    header.set_cksum();
+    dst.write_all(header.as_bytes())?;
+
+    Ok(())
+}
+
 fn append_dir_all(
     dst: &mut dyn Write,
     path: &Path,
     src_path: &Path,
     mode: HeaderMode,
     follow: bool,
 ) -> io::Result<()> {
-    let mut stack = vec![(src_path.to_path_buf(), true, false)];
-    while let Some((src, is_dir, is_symlink)) = stack.pop() {
+    let metadata = fs::metadata(src_path)?;
+    if !metadata.is_dir() {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!("{} is not a directory", src_path.display()),
+        ));
+    }
+    let mut stack = vec![(src_path.to_path_buf(), metadata)];
+    while let Some((src, metadata)) = stack.pop() {
         let dest = path.join(src.strip_prefix(&src_path).unwrap());
+        let file_type = metadata.file_type();
+        let is_dir = file_type.is_dir();
+        let is_symlink = file_type.is_symlink();
         // In case of a symlink pointing to a directory, is_dir is false, but src.is_dir() will return true
         if is_dir || (is_symlink && follow && src.is_dir()) {
             for entry in fs::read_dir(&src)? {
                 let entry = entry?;
-                let file_type = entry.file_type()?;
-                stack.push((entry.path(), file_type.is_dir(), file_type.is_symlink()));
+                let metadata = entry.metadata()?;
```

Comment:
This unfortunately has a side effect where on platforms like Unix this ends up doing a syscall whereas previously accessing `file_type` generally doesn't require a syscall on most major platforms. The intention was that this loop should primarily only do `fs::read_dir` as a sycall and otherwise other syscalls can be avoided.

My thinking about refactoring this was largely that this loop has one iteration of "look at the file type and delegate appropriately" but above in `append_path_with_name` it's basically the same loop. With the `append_path_with_name` addition it's getting more complicated so I was wondering if it would be possible to refactor this loop and `append_path_with_name` above to call a common function which is the one location where special files are handled.

Reply:
> This unfortunately has a side effect where on platforms like Unix this ends up doing a syscall whereas previously accessing `file_type` generally doesn't require a syscall on most major platforms. The intention was that this loop should primarily only do `fs::read_dir` as a sycall and otherwise other syscalls can be avoided.
> 

I have reverted the changes made to this loop.

> My thinking about refactoring this was largely that this loop has one iteration of "look at the file type and delegate appropriately" but above in `append_path_with_name` it's basically the same loop. With the `append_path_with_name` addition it's getting more complicated so I was wondering if it would be possible to refactor this loop and `append_path_with_name` above to call a common function which is the one location where special files are handled.

I am not quite sure which one is better:

1. Make a new delegate function `append_special` and rename the current `append_special` to `append_special_unix`. Then in the `append_special` function, use `fs::metadata` to get the metadata and pass it to `append_special_unix`. In which case, the loop needs to be modified to differentiate a regular file from a special one.
2. Similar to (1) but instead of using `fs::metadata` in that function, `fs::metadata` is used in the "special file" branch of the loop. The advantage of this is one less `stat` when called from `append_path_with_name` (since there is one `stat` in that function as well).
3. Similar to (1) but making `append_special` to accept `Option<fs::metadata>` and `stat` on-demand.